plugins/ltac/tacinterp.ml

```diff
@@ -1610,10 +1610,10 @@ and interp_atomic ist tac : unit Proofview.tactic =
             (k,(make ?loc f))) cb
         in
         let sigma,tac = match cl with
-          | None -> sigma, Tactics.apply_with_delayed_bindings_gen a ev l
-          | Some cl ->
-              let sigma,(id,cl) = interp_in_hyp_as ist env sigma cl in
-              sigma, Tactics.apply_delayed_in a ev id l cl in
+          | [] -> sigma, Tactics.apply_with_delayed_bindings_gen a ev l
+          | cl ->
+              let sigma,cl = List.fold_left_map (interp_in_hyp_as ist env) sigma cl in
+              sigma, Tacticals.New.tclMAP (fun (id,cl) -> Tactics.apply_delayed_in a ev id l cl) cl in
```

Comment:
It's not clear that this is the desired semantics. If I read the implementation correctly, this means that if one `apply` generates several subgoals, then we'll perform the remaining `apply` on all subgoals. This is different from the semantics of `apply foo, bar` for instance.

Reply:
Good remark. We indeed expect it to be chained only on the main subgoals (i.e. the one produced by the `as`), and not on the side conditions. I.e., in:
```
Goal forall f : unit -> nat -> bool, nat -> nat -> nat.
intros f n m.
apply f in n as [], m as [].
```
we would expect that it skips the side condition (coming last) and behaves like:
```
apply f in n as []; [ apply f in m as [] .. | ].
```
Not clear how to do it. I could for instance extend the passing-continuation-passing style currently used to chain `apply l1, l2 in id`. Unless someone has a better idea.

Reply:
I used continuation-passing then. Are you ok with the new semantics?

Reply:
The semantics is OK, but the use of the CPS is a bit ugly. It might be the right time to define a combinator similar to `enter` that only applies to the "principal" goal, if there is one. `apply` does mark one of its output goals as principal, right?

Reply:
Forget what I just said, the combinator idea is broken. I think we could still do better though, maybe returning the set of "principal" apply goals instead of threading the tactic. It's probably going to require more code change, but it'd be definitely better for future modifications.

Reply:
> Forget what I just said, the combinator idea is broken. I think we could still do better though, maybe returning the set of "principal" apply goals instead of threading the tactic. It's probably going to require more code change, but it'd be definitely better for future modifications.

Sound like a nice idea. So, some refinement of `principal_future_goal` in `evar_map` which supports several main future goals + a combinator of type say `tclTHENFOCUS :   focus tactic -> unit tactic -> unit tactic` + a combinator `tclUNITFOCUS : focus tactic` with `focus = goal list` or something like that???

In any case, this shall be for after 8.2 then.

Reply:
No, you can't store that in the evarmap otherwise you'll get back to the model pre-8.5. I was rather advocating that `apply` should have type `goal list tactic` or something, where the returned list contains the "principal goals", whatever that means in the context of `apply`. I think that precisely the teachings of the new semantics is that "principal goal" is not a universal notion, we only use it inside `refine` because it's somewhat hardcoded in the primitive.